packages/ResizeObserver/src/ResizeObserver.js

```diff
@@ -0,0 +1,95 @@
+import React from "react";
+import PropTypes from "prop-types";
+import useResizeObserver from "use-resize-observer/polyfilled";
+import debounce from "lodash.debounce";
+import { ShirtSizes } from "@paprika/helpers/lib/customPropTypes";
+import * as sc from "./ResizeObserver.styles";
+
+const propTypes = {
+  breakpointLarge: PropTypes.number,
+  breakpointSmall: PropTypes.number,
+  children: PropTypes.node,
+  debounceDelay: PropTypes.number,
+  isFullWidth: PropTypes.bool,
+  isFullHeight: PropTypes.bool,
+  onBreak: PropTypes.func,
+  onResize: PropTypes.func,
+};
+
+const defaultProps = {
+  breakpointLarge: 768,
+  breakpointSmall: 360,
+  children: null,
+  debounceDelay: 200,
+  isFullWidth: true,
+  isFullHeight: false,
+  onBreak: () => {},
+  onResize: () => {},
+};
+
+const MAX_WAIT = 1000;
+
+const ResizeContext = React.createContext();
+
+function useObservedDimensions() {
+  const { width, height } = React.useContext(ResizeContext);
+  return { width, height };
+}
+
+function useBreakpoints() {
+  const { size } = React.useContext(ResizeContext);
+  return { size };
+}
+
+function getSize(width, breakpointSmall, breakpointLarge) {
+  let size = ShirtSizes.MEDIUM;
+  if (breakpointSmall && width <= breakpointSmall) {
+    size = ShirtSizes.SMALL;
+  } else if (breakpointLarge && width >= breakpointLarge) {
+    size = ShirtSizes.LARGE;
+  }
+  return size;
+}
+
+function ResizeObserver(props) {
+  const { breakpointSmall, breakpointLarge, children, debounceDelay, onBreak, onResize, ...moreProps } = props;
+  const refContainer = React.useRef(null);
+  const [{ width, height }, setDimensions] = React.useState({});
+  const [size, setSize] = React.useState(null);
+
+  function handleResize({ width, height }) {
+    setDimensions({ width, height });
+    onResize({ width, height });
+
+    const newSize = getSize(width, breakpointSmall, breakpointLarge);
+    if (newSize !== size) {
+      setSize(newSize);
+      onBreak(newSize);
+    }
+  }
+
+  useResizeObserver({
+    ref: refContainer,
+    onResize: debounce(handleResize, debounceDelay, { maxWait: MAX_WAIT }),
+  });
+
+  React.useLayoutEffect(() => {
+    const { width, height } = refContainer.current.getBoundingClientRect();
+    setDimensions({ width, height });
+    setSize(getSize(width, breakpointSmall, breakpointLarge));
+  }, [children, breakpointSmall, breakpointLarge]);
+
+  return (
+    <sc.ResizeObserver data-pka-anchor="resize-observer" {...moreProps} ref={refContainer}>
+      <ResizeContext.Provider value={{ width, height, size }}>{children}</ResizeContext.Provider>
+    </sc.ResizeObserver>
+  );
+}
+
+ResizeObserver.displayName = "ResizeObserver";
+ResizeObserver.propTypes = propTypes;
+ResizeObserver.defaultProps = defaultProps;
+ResizeObserver.useObservedDimensions = useObservedDimensions;
+ResizeObserver.useBreakpoints = useBreakpoints;
```

Comment:
Not sure about this convention of attaching the custom hooks I want to export right to the component... I like it, but can't think of an example where I've seen it before.

Reply:
We use this pattern usually i believe.

`import { useColumnsArrangement } from '@paprika/action-bar'`

Any reason for you to not do it this way instead?

Reply:
Well, it means my import in the consumer is a little neater, I can just do 
`import ResizeObserver  from '@paprika/action-bar';`  

And I can export from `ResizeObserver.js` instead of from `index.js` (which I don't usually look at).  

Neither are deal-breakers, but since both approaches work, I kind of prefer the way I'm proposing here.

Reply:
React suggest that custom hooks should always start with use, I guess adding the `ResizeObverser` kind of break this principle, the reason they suggested is because helps static analyzers to do a proper job on the hooks. However, In this case, I know there is nothing to analyze but somehow is not a regular way of exporting hooks.

That's why I end up exporting the hook for the Action bar as a property of the import instead of adding it directly to the ActionBar as you are proposing. 

I think while is not more beautiful, is more "standard" <- whatever that means 

> Custom Hooks are more of a convention than a feature. If a function’s name starts with ”use” and it calls other Hooks, we say it is a custom Hook. The useSomething naming convention is how our linter plugin is able to find bugs in the code using Hooks.
https://reactjs.org/docs/hooks-overview.html